### ISSUE TITLE
Using a non-balanced { in a mapping or execute-keys inside a hook

### ISSUE THREAD

**Issue by aidam38** [mawww/kakoune#2760]
While setting up my LaTeX hooks, I encountered a weird problem, where the `{` and `}` characters inside a snippet interfere with the `{` and `}` characters used for expansions in kakoune configuration. It only happens when the mapping is inside a hook. Here, I provide a Minimal Working Example:
```
hook global WinSetOption filetype=latex %{ 
    map global user e 'i\{'
}
```
The error I get is
```
kakrc_test:parse error: 0:41: unterminated string '%{...}'
Error: 1:1: 'source' parse error: 0:41: unterminated string '%{...}'
```

**Comment by eraserhd** [comment]
The `%{ ... }` quoting syntax, which you are using here, requires inside braces to be balanced.  You could use `%( ... )` or `%< ... >` instead, or you can add a comment at the end of the map line like so:

```
hook global WinSetOption filetype=latex %{ 
    map global user e 'i\{' # } <-- balance braces
}
```

Kakoune's configuration language is much like shell, although a bit saner because of the ability to nest.  Kakoune doesn't know the inside brace is quoted at the time it's collecting the hook commands because it doesn't know yet that it is looking at more configuration language.

**Comment by andreyorst** [comment]
> or you can add a comment at the end of the map line like so

Another workaround:
```
hook global WinSetOption filetype=latex %{ evaluate-commands %sh{
    opener='{'; closer='}'
    printf "%s\n" "map global user e 'i${opener}'"
}}
```

**Comment by mawww** [comment]
You can also use non nestable punctionation characters, which can be escaped inside the string by doubling them up:
```
hook global WinSetOption filetype=latex %~ evaluate-commands %sh`
    printf "%s\n" "map global user e 'i{'"
`~
# or (but a bit confusing)
hook global WinSetOption filetype=latex %~ evaluate-commands %sh~~
    printf "%s\n" "map global user e 'i{'"
~~~
```

**Comment by alexherbo2** [comment]
How about a doubling syntax?

``` kak
evalutate-commands %{{
  echo '{'
}}
```

**Comment by eraserhd** [comment]
On Tue, Mar 12, 2019 at 10:23 PM Alex Leferry 2> <notifications@github.com>
wrote:

> How about a doubling syntax?
>
> evalutate-commands %{{
>   echo '{'
> }}
>
>
That would prevent quoting arguments with surrounding braces.

**Comment by andreyorst** [comment]
In rust there's such construction, called raw string, that allows to easily include quotes inside strings:
```rust
let vaiv = r#"
    some = "strings"
    inside = "the main string"
"#;
```

If we want to include `"#", we simply double up starting and end quotes:
```rust
let vaiv = r##" "# "##;
let vaiv = r###" "## "###;
// ...
let vaiv = r########" "####### "########;
```

I guess, that something similar could be used in Kakoune, as this mechanism of escaping is one of the best that I've seen.

**Comment by pickfire** [comment]
Nah, the best method of escaping is to use field separator, group separator, record separator, unit separator that would never be in the code. https://stackoverflow.com/a/29811033

```
1c  FS  ␜  ^\  File Separator
1d  GS  ␝  ^]  Group Separator
1e  RS  ␞  ^^  Record Separator
1f  US  ␟  ^_  Unit Separator
```
